components/psst/browser/content/psst_tab_helper.cc

```diff
@@ -0,0 +1,128 @@
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/thread_pool.h"
+#include "base/values.h"
+#include "brave/components/psst/browser/core/psst_rule.h"
+#include "brave/components/psst/browser/core/psst_rule_service.h"
+#include "brave/components/psst/common/features.h"
+#include "components/sessions/content/session_tab_helper.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+
+namespace psst {
+
+// static
+void PsstTabHelper::MaybeCreateForWebContents(content::WebContents* contents,
+                                              const int32_t world_id) {
+  if (contents->GetBrowserContext()->IsOffTheRecord() ||
+      !base::FeatureList::IsEnabled(psst::features::kBravePsst)) {
+    return;
+  }
+
+  psst::PsstTabHelper::CreateForWebContents(contents, world_id);
+}
+
+PsstTabHelper::PsstTabHelper(content::WebContents* web_contents,
+                             const int32_t world_id)
+    : WebContentsObserver(web_contents),
+      content::WebContentsUserData<PsstTabHelper>(*web_contents),
+      world_id_(world_id),
+      weak_factory_(this) {
+  psst_service_ = PsstRuleService::GetInstance();
+  if (!psst_service_) {
+    return;
+  }
+}
+
+PsstTabHelper::~PsstTabHelper() = default;
+
+void InsertScriptInPage(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    const std::string& script,
+    content::RenderFrameHost::JavaScriptResultCallback cb) {
+  content::RenderFrameHost* render_frame_host =
+      content::RenderFrameHost::FromID(render_frame_host_id);
+
+  if (render_frame_host) {
+    render_frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(script), std::move(cb), world_id);
+  }
+}
+
+void InsertPolicyScript(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    std::string policy_script) {
+  InsertScriptInPage(render_frame_host_id, world_id, policy_script,
+                     base::DoNothing());
+}
+
+void OnTestScriptResult(
+    std::string policy_script,
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    base::Value value) {
+  if (value.GetIfBool().value_or(false)) {
+    InsertPolicyScript(render_frame_host_id, world_id, policy_script);
+  }
+}
+
+void InsertTestScript(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    MatchedRule rule) {
+  CHECK(render_frame_host_id);
+
+  InsertScriptInPage(render_frame_host_id, world_id, rule.test_script,
+                     base::BindOnce(&OnTestScriptResult, rule.policy_script,
+                                    render_frame_host_id, world_id));
+}
+
+void PsstTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  CHECK(navigation_handle);
```

Comment:
nit: this really isn't necessary here and doesn't really add any value. If navigation_handle is null then there is a serious upstream bug that would be affecting every tab helper